### PR TITLE
feat(diff): revert individual hunks from the diff view #13

### PR DESCRIPTION
**Pull request by Blankeos**
Adds a per-hunk revert affordance on unstaged file diffs. In the Files
panel, <c-j>/<c-k> cycle through hunks, <enter> reverts the selected
hunk, and clicking the marker glyph in the divider column reverts that
hunk directly. Hunks are reverted via `git apply --reverse` against
the worktree, scoped to one hunk at a time.

---

Hi @fcmiranda  I made a smaller-scoped PR with your changes + a few bug fixes of my own, still crediting your commit.

I'll also try to do a rustfmt in a later commit in main (this is my bad for not doing that to begin with).

https://github.com/user-attachments/assets/e9195c81-e1d0-41f3-b0db-d1b38ce79d56

Smaller scoped version of https://github.com/Blankeos/lazygitrs/pull/12

### CODE REVIEW

File: src/pager/side_by_side.rs [hunk 1]
```diff
@@ -530,6 +546,11 @@ impl DiffViewState {
         self.hunk_line_offsets = parsed.hunk_line_offsets;
         self.sections = parsed.sections;
         self.file_exists_on_disk = parsed.file_exists_on_disk;
+        self.selected_revert_hunk = if same_file {
+            prev_selected_revert_hunk.filter(|&i| i < self.hunk_starts.len())
+        } else {
+            None
+        };
```
[the remark this answers is not in the document]
Comment (Blankeos) [reply src/pager/side_by_side.rs:598]: This is a real bug. But not the fix.. Currently fixing it.

**Comment by Blankeos** [comment]
Also added undo btw, it's on `u`.

I'm debating whether to keep `c-j/c-k`. I think we could just use `{/}` for cycling. And then maybe use ctrl-z or opt-z for the revert.

**Comment by fcmiranda** [comment]
Nice! Thanks for crediting the commit.

What do you think also add the stage button? If yes, any idea for a keybinding to cycling it?

image


image

**Comment by fcmiranda** [comment]
@Blankeos I've tested your pr and unfortunately the navigation is not working with ctrl k/ ctrl j, but I think was intentional right?

I'll fork your branch and open your PR

**Comment by Blankeos** [comment]
HI fcmiranda! Sorry, was away in HK this week. Anyway, I have made some changes I decided to keep this feature simple!

1. No more `c-k` and `c-j` to cycle just do the regular `shift+]` and `shift+[` to navigate through hunks.
2. Press `enter` to show a bunch of actions i.e. "Revert hunk".
3. Press `u` to undo (it's also shown as a tip at the bottom-right that you can undo a revert).

> What do you think also add the stage button

The stage button is a really good idea, good feedback. Thanks! I'll think about it for a bit. But I think I should not block this PR anymore. Let's get it released w/ a bunch more important bug fixes I did on main.

**Comment by Blankeos** [comment]
# PR Review — #13 `feat(diff): revert individual hunks from the diff view` (claude)

**Branch:** `feature/revert-block-clean` → `main`
**Merge confidence:** **4 / 5** ✅

## Short body (paste-into-GitHub)

Solid, well-scoped feature. The `--unidiff-zero` carved sub-patch approach is the right call — it correctly handles visual blocks narrower than the surrounding `@@` hunk, pure insertions/deletions, and the `\ No newline at end of file` marker. Undo via snapshot is bounded (`REVERT_UNDO_STACK_CAP = 20`), failure modes route to popups, the menu opens with **Cancel** focused so a stray Enter does nothing destructive, and the status-bar/footnote/hover affordances are well thought out. Tests pass, no new warnings on touched files, no migrations needed (`#[serde(default)]` on the keybinding struct means existing user configs just inherit the new defaults). Only real blocker is the stray `- [ ] 123` entry in `TODOS.md`. PR body is also slightly out of sync with the final keymap.

---

## What changed

| File | Δ | Notes |
|---|---|---|
| `src/git/staging.rs` | +144 / -0 | Adds `revert_visual_block_in_worktree` and the `build_visual_block_patch` carver |
| `src/gui/mod.rs` | +218 / -41 | Keybindings, mouse hover/click, hunk context menu, undo plumbing |
| `src/pager/side_by_side.rs` | +625 / -76 | Marker glyph in divider column, tooltip, undo stack/state, row-mapping helpers (`line_chunk_at_row`, `line_visual_height`), wrap-mode row counting |
| `src/gui/views.rs` | +72 / -34 | Status-bar split: diff-focused hint set vs sidebar hints; revert-marker plumbing |
| `src/config/keybindings.rs` | +6 / -0 | New `revertBlock`/`undoRevertBlock` fields |
| `src/gui/presentation/diff_mode.rs` | +1 / -1 | Pass `show_revert_markers: false` for diff-mode renderer |
| `README.md` | -9 | Removes test markers (1/2/2.1/3/last/1) |
| `TODOS.md` | +2 | ⚠ One real item + one stray `- [ ] 123` |

## Strengths

- **Right git primitive.** `git apply --reverse --unidiff-zero` on a freshly built per-block patch — no fragile string surgery against the user's file.
- **Visual block ≠ `@@` hunk.** The carver in `build_visual_block_patch` walks the unified diff line-by-line and only emits lines whose old/new line numbers fall in the requested ranges, so a single `@@` containing two visual blocks separated by a few context lines is split correctly. Pure inserts/deletes (`want_old` or `want_new` is `None`) and `\ No newline at end of file` markers are handled explicitly (`staging.rs:131-140`).
- **Bounded undo.** Snapshots capped at 20 (`REVERT_UNDO_STACK_CAP`), with a high-water-mark denominator that resets when the stack drains — clean UX detail.
- **Safe menu default.** `Cancel` is the first item in the hunk popup, so Enter-on-Enter doesn't revert (`gui/mod.rs:5101-5111`).
- **Survives same-file reloads.** `apply_parsed`/`load`/`apply_diff_output` all preserve `selected_revert_hunk`/`hovered_revert_hunk` only when the filename matches, and re-clamp to current `hunk_starts.len()`.
- **No config migration.** `UniversalKeybinding` already has `#[serde(default)]` (`keybindings.rs:32`), so users with old configs just get the new defaults silently.
- **Marker is scoped.** `show_revert_markers` only true in the Files context (`views.rs:101, 588`); diff-mode and commit/stash diffs unchanged.
- **Discoverability.** Help text updated for both Files panel and diff-focused dialog; status bar shows `{/}` + `enter` (emphasized when a hunk is selected) + `u` (only when there's something to undo); bottom-right footnote shows `u undo revert (n/m)`.

## Concerns / nits

1. **🟡 Stray TODO entry.** `TODOS.md` has a `- [ ] 123` line. Remove before merge.
2. **🟡 PR description out of sync.** Body says `<c-j>/<c-k> cycle through hunks, <enter> reverts the selected hunk`, but the merged keymap is `{/}` (the commit `dbd6941` renamed this) and `<enter>` opens a confirmation menu rather than reverting directly. Update the description so reviewers and future-you don't get confused.
3. **🟡 Local-only README leftovers.** `README.md` has unstaged changes (`1, 2, 2.1, 3, last, 1` markers) — these aren't on the PR yet. Clean up + `just sync_readme` (per `CLAUDE.md`) before merging.
4. **🟢 Formatting churn.** `side_by_side.rs` has ~150 lines of cosmetic rustfmt re-wrapping mixed into the substantive changes. PR body already acknowledges this ("I'll also try to do a rustfmt in a later commit") — fine, just makes the review slower than it needs to be. Future PRs: land formatting separately.
5. **🟢 Enter-in-diff-focus is universally swallowed.** `handle_diff_focused_key` returns `Ok(())` for any Enter press regardless of `ContextId`. Functionally identical to the prior `_ => {}` no-op, but worth noting that if you later want Enter to do something else in diff-focus from a non-Files context, the early-return guards will need narrowing.
6. **🟢 "u" key choice.** Doesn't collide with the global `undo` (`z`) or any other diff-focus binding. Universally scoped but no-ops outside Files. Fine — just noting the choice for future maintenance.

## Migrations / pre-merge checklist

**Migrations:** None.
- No DB, no schema, no on-disk state migration.
- Config: new fields have `#[serde(default)]` defaults; old `config.yml` files load unchanged.
- Undo stack lives only in process memory (intentional — comment in `gui/mod.rs:5181-5183` makes this explicit).

**Before merging:**
- [x] Remove `- [ ] 123` from `TODOS.md`
- [x] Resolve local `README.md` cleanup + run `just sync_readme` (so `npm/README.md` stays in sync)
- [x] Update PR description: `{/}` (not `<c-j>/<c-k>`); `<enter>` opens hunk menu (not direct revert)
- [x] Smoke-test the four edge cases the carver explicitly handles:
  - Hunk with both `-` and `+` (modified block)
  - Pure-insert hunk (no `-` lines)
  - Pure-delete hunk (no `+` lines)
  - File ending without trailing newline (`\ No newline at end of file`)
- [x] (Optional) One file with two visual blocks inside the same `@@` separated by 1–3 context lines, to confirm only the targeted block reverts

## Local verification I ran

- `cargo check` — 0 errors, 46 pre-existing warnings (no new warnings introduced by this PR's files).
- `cargo test` — 6 passed / 0 failed.
- `git diff main...HEAD --stat` — 8 files, +1110 / -119.
- Inspected key paths: revert handler scoping, config defaults, undo stack mechanics, hover/selected/click state machine, marker render conditions, edge-case handling in `build_visual_block_patch`.
